File: python/maths/Peterson_number.py
# This is an example for peterson number.
#  145 = 5! + 4! + 1!
#      = 120 + 24 +1
#      = 145


#This list contains factorials of all the nine digits in order.
#This is beacause, we may need to compute these factorials many times in the numbers given.

factorials = [1, 1, 2, 6, 24, 120, 720,5040, 40320, 362880]

#Function that checks the peterson number condition
def peterson(n):
    num = n
    sum = 0

    while n > 0:
        digit = int(n % 10)
        sum += factorials[digit]
        n = int(n / 10)
 
    # Condition check for a number
    # to be a Peterson Number
    if(sum==num):
        return True
    else:
        return False
 
 
# Driver Code
number = input("Enter the number to be checked for peterson's condition : ");
print("Yes, the number {} is peterson number".format(number) if peterson(int(number)) else "No, {} is not peterson number".format(number))

File: cpp/maths/Peterson_number.cpp
//A number is said to be a Peterson number if the sum of factorials of each digit 
//is equal to the number itself

//Given a number n, check if it is a Peterson number

#include <bits/stdc++.h>

using namespace std;

//function to get factorial of a number
int fact(int n) {
	if(n == 0) {
		return 1;
	}
	return n * fact(n-1);
}

//check if a number is a Peterson number
bool isPeterson(int n) {

	//convert the number into string to easily traverse over its digits
	string s = to_string(n);

	int sumOfFact = 0;

	//find sum of factorial of digits of the number
	for(char ch : s) {
		sumOfFact += fact(ch-'0');
	}

	//return true is sum of factorial of digits is equal to the number itself
	return sumOfFact == n;
}

int main() {

	//take the input as a number
	cout << "Enter the number:\n";
	int n;
	cin >> n;

	//call the function to check if this is a Peterson number or not
	if(isPeterson(n)) {
		cout << "Yes, this is a Peterson number\n";
	} else {
		cout << "No, this is not a Peterson number\n";
	}
	return 0;
}